### ISSUE TITLE
How to reload Infinite Hits + filters when clicking back in browser

### ISSUE THREAD

**Issue by alexpchin** [algolia/react-instantsearch#2971]
I've currently got gatsby website a `react-instantsearch-dom` set of results with `connectInfiniteHits`. There are a couple of filters being applied to the results using a few. `connectRefinementList` connectors.  The corresponding results are displayed using a masonry layout.

When I click an item, I want to navigate to the detail page for the image. This is all working. However, when I click back, I want to display the page as it was before navigating to the detail page.

I'm thinking that I need to store the hits in redux and load from redux? Do you have an example of this? Or any suggestions on how to achieve this? I can't seem to find a solid example.

Other similar questions:

- https://github.com/algolia/react-instantsearch/issues/1061
- https://github.com/algolia/react-instantsearch/issues/85
- https://github.com/algolia/react-instantsearch/issues/1170
- https://github.com/algolia/instantsearch.js/issues/2835

**Comment by Haroenv** [comment]
We have a new feature called `infiniteHitsCache` for this: https://github.com/algolia/react-instantsearch/pull/2921. Does that fit what you are trying to achieve?

**Comment by alexpchin** [comment]
Hi @Haroenv, 

I'm still struggling a little bit actually :( 

I'm using the connector `connectInfiniteHits` rather than `InfiniteHits`. At the moment, I am using a hook `useIntersect` to call `refineNext` automatically.

I'm using `@reach/router` as I am using Gatsby.

If I make the hitsPerPage a low number, I get really inconsistent results displaying. Any ideas? 

Thanks for your help!

**Comment by alexpchin** [comment]
It's like the hits load the last "batch" that was loaded but not load from the beginning?

**Comment by alexpchin** [comment]
This may help? This is where I am at. https://production.iynk.com/tattoos

To reproduce an entire site in a codesandbox would be too much but you might be able to understand the issue I have?

**Comment by Haroenv** [comment]
- if `loadPrevious` is true, it allows for a button to load the previous page of results
- if `cache` is set to e.g. `createInfiniteHitsSessionStorageCache`, it allows the previously shown hits to be remembered

This is indeed also the case if you're using the connector, but the behaviour might not be as clear if the hits aren't shown in a flat, numbered list.

does cache work for you? If so that's something we might want to document in more detail than here: https://www.algolia.com/doc/api-reference/widgets/infinite-hits/react/#widget-param-cache

**Comment by alexpchin** [comment]
Hi @Haroenv,  I have set the cache as per that example. Let's say you load 60 results, e.g. 3 x 20 results and you go to a detail page, and return, should the cache load 60 results immediately from the cache? It seems like the page loads the results from the correct page but doesn't retain the previous results?

**Comment by Haroenv** [comment]
yes, it should retain the previous results, otherwise that's a bug. Do you have a reproduction in code sandbox?

**Comment by alexpchin** [comment]
Hey, @Haroenv I made a code sandbox here: https://codesandbox.io/s/lucid-dubinsky-114jg When using the InfiniteHiits it works but not in the implementation that I have outlined here. Any help would be appreciated!

**Comment by Haroenv** [comment]
I ran that page and the behaviour seemed to be correct, if I reloaded it still showed items 0-100 like I scrolled. Do you think you could make e.g. a screen recording of it behaving wrong? 

Since you're using intersection observer, is it possible that during a loading state, where the hits aren't shown yet from the cache, you already render the sentinel, which causes it to change (or skip) pages? Make sure you call that only once and not multiple times for the same page

**Comment by alexpchin** [comment]
Ok, I've started to nail down the problem. You are correct, the implementation I have previously created in the codesandbox does work as expected... The issue comes from when you include a `<Configure>` in either a single Index or multi-index search.

In my app, I have a custom `<Filter />` component that is re-used between pages. In this Filter, I render a list of tags from a different index based on the search result, it uses a `<Configure>` to change the number of results displayed.

Initially, I thought it was to do with adding a second `<Index>` but I think it's actually when you include a `<Configure>`.

I've updated the example. Any thoughts on a solution would be appreciated. https://codesandbox.io/s/lucid-dubinsky-114jg

**Comment by alexpchin** [comment]
@Haroenv What do you think about allowing the ability to pass in a custom key to allow caches for different lists?

```js
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties"
import isEqual from "react-fast-compare"

function getStateWithoutPage(state) {
  var _ref = state || {},
    page = _ref.page,
    rest = _objectWithoutProperties(_ref, ["page"])

  return rest
}

function hasSessionStorage() {
  return (
    typeof window !== "undefined" &&
    typeof window.sessionStorage !== "undefined"
  )
}

export default function createInfiniteHitsSessionStorageCache(
  KEY = "ais.infiniteHits"
) {
  return {
    read: function read(_ref2) {
      var state = _ref2.state

      if (!hasSessionStorage()) {
        return null
      }

      try {
        var cache = JSON.parse(window.sessionStorage.getItem(KEY))
        return cache && isEqual(cache.state, getStateWithoutPage(state))
          ? cache.hits
          : null
      } catch (error) {
        if (error instanceof SyntaxError) {
          try {
            window.sessionStorage.removeItem(KEY)
          } catch (err) {
            // do nothing
          }
        }

        return null
      }
    },
    write: function write(_ref3) {
      var state = _ref3.state,
        hits = _ref3.hits

      if (!hasSessionStorage()) {
        return
      }

      try {
        window.sessionStorage.setItem(
          KEY,
          JSON.stringify({
            state: getStateWithoutPage(state),
            hits: hits,
          })
        )
      } catch (error) {
        // do nothing
      }
    },
  }
}

```

**Comment by alexpchin** [comment]
@Haroenv Did you have any ideas why including a `<Configure>` resets the cache as per https://github.com/algolia/react-instantsearch/issues/2971#issuecomment-696626736 Thanks in advance.

**Comment by carlosbaraza** [comment]
@Haroenv do you know why the cache resets when using the `<Configure>` component?

**Comment by carlosbaraza** [comment]
I think the problem with using `Configure` is that when coming back to the search results, on the first render, the `searchState` does not include the configuration added with `Configure`. The problem [is on this line](https://github.com/algolia/react-instantsearch/blob/master/packages/react-instantsearch-core/src/connectors/connectInfiniteHits.js#L84).

The solution I am using for the moment is to initialize the `searchState` with the configuration added to `Configure`.

```jsx
  const [searchState, setSearchState] = useState({
    ...props.searchState,
    configure: {
      clickAnalytics: !skipAnalytics,
    },
  });

...

      <InstantSearch
        searchState={searchState}
        ...
      >
        <Configure clickAnalytics />
      ...
```

**Comment by rigdal** [comment]
I am also getting this. Works on refresh but not on back button.

**Comment by rashad** [comment]
Hi everyone,

First a big s/o to everyone contributing to the library it's a real masterpiece! I'm here because I'm willing to use the `createInfiniteHitsSessionStorageCache` object but I can't find any documentation on how to use it with a custom infinit hits component built with `connectInfiniteHits`. 

I got that I could pass the `createInfiniteHitsSessionStorageCache` as a prop to the regular `InfinitHits` components but I can't find a way of using it with a custom component. Any chance someone could help me?

Cheers from 🇫🇷 👋

**Comment by Haroenv** [comment]
you can pass it to the custom infinite hits the same as to the builtin one, as it's the connector that does something with the cache prop @rashad. If you're stuck somewhere, please create a sandbox. We have a template available for that purpose [here](https://codesandbox.io/s/github/algolia/create-instantsearch-app/tree/templates/react-instantsearch). Thanks!

**Comment by rashad** [comment]
Hey @Haroenv ! Thanks a lot for the info. I've been digging ever since my issue and I still haven't been able to make it work. Could you help me understand why [the sandbox provided](https://codesandbox.io/s/xenodochial-mcnulty-o7qdv?file=/src/App.js) in the [feature PR](https://github.com/algolia/react-instantsearch/pull/2921) seems to be broken as well ? 

I've arrived to a point where the page parameter is ok in my URL and safely transmitted as an integer into the search state. But somehow, the scroll does not happen.

Looking forward to your answer 👋

**Comment by rashad** [comment]
Hello there, quick update I finally made it work 🎉 

The bug seemed to be lying in the `autoFocus` prop I was passing from the `<CustomSearchBox />` to the actual `<input/>` (I've removed it). As soon as I passed it to the input I got issues with the cache.

Side note as well, it seems that TypeScript has an issue with the `cache` prop I'm passing since I'm getting this error
```ts
 Property 'cache' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<Component<Omit<InfinitHitsProps, InfiniteHitsProvided<unknown>>, any, any>> & Readonly<...> & Readonly<...>'.
```
knowing that `InfinitHitsProps` interface extends `InfiniteHitsProvided`. I'll try to disable it unless anyone has a better solution ! 

Cheers ! 👋 🇫🇷

**Comment by Fyvel** [comment]
Hello hello,
I ran into the same Typescript issue, rather than disabling it you can tweak a bit the props type
```ts 
type Props<T> = InfiniteHitsProvided<T> & { cache: any }
function CustomHits({ hits, hasPrevious, refinePrevious, hasMore, refineNext }: Props<SearchResult>) {
  // ...
}

export const HitsConnector = connectInfiniteHits(CustomHits)
```

**Comment by atrickofthetail** [comment]
> Hello there, quick update I finally made it work 🎉
> 
> The bug seemed to be lying in the `autoFocus` prop I was passing from the `<CustomSearchBox />` to the actual `<input/>` (I've removed it). As soon as I passed it to the input I got issues with the cache.
> 
> Side note as well, it seems that TypeScript has an issue with the `cache` prop I'm passing since I'm getting this error
> 
> ```ts
>  Property 'cache' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<Component<Omit<InfinitHitsProps, InfiniteHitsProvided<unknown>>, any, any>> & Readonly<...> & Readonly<...>'.
> ```
> 
> knowing that `InfinitHitsProps` interface extends `InfiniteHitsProvided`. I'll try to disable it unless anyone has a better solution !
> 
> Cheers ! 👋 🇫🇷

Any ideas why the `autoFocus` is causing this? I am running into the same issue, where I am setting it directly on an `<input />` and with the prop set, cache is reset navigating to another page and back. Without the `autoFocus` prop, it works just fine,

**Comment by dhayab** [comment]
Hi! We're doing a round of cleanup before migrating this repository to the new [InstantSearch monorepo](https://github.com/algolia/instantsearch.js).

This issue seems not to have generated much activity lately, so we're going to close it as it has multiple implementation answers. Feel free to open a new issue if necessary.
